.ci/magician/cmd/generate_comment.go

```diff
@@ -0,0 +1,374 @@
+			diffProccessorErr = err
+		}
+	}
+	for _, path := range []string{"old", "new", "bin"} {
+		if err := r.RemoveAll(filepath.Join(diffProcessorPath, path)); err != nil {
+			fmt.Printf("error removing files: %v\n", err)
+			os.Exit(1)
+		}
+	}
+	for _, path := range []string{"old", "new"} {
+		if err := r.Copy(tpgbLocalPath, filepath.Join(diffProcessorPath, path)); err != nil {
+			fmt.Printf("Error copying files: %v\n", err)
+			os.Exit(1)
+		}
+	}
+
+	if diffProccessorErr != nil {
+		fmt.Println("Breaking changes failed")
+		breakingChanges = `## Breaking Change Detection Failed
+The breaking change detector crashed during execution. This is usually due to the downstream provider(s) failing to compile. Please investigate or follow up with your reviewer.`
+	} else {
+		fmt.Println("Breaking changes succeeded")
+		breakingChanges = compareBreakingChanges(tpgBreaking, tpgbBreaking)
+	}
+
+	tpgbLocalPathOld := tpgbLocalPath + "old"
+	if err := r.Copy(tpgbLocalPath, tpgbLocalPathOld); err != nil {
+		fmt.Printf("Error copying files from %s to %s: %v\n", tpgbLocalPath, tpgbLocalPathOld, err)
+		os.Exit(1)
+	}
+	if stdout, stderr, err := r.Run(tpgbLocalPathOld, "git", []string{"checkout", "origin/" + oldBranch}, nil); err != nil {
+		fmt.Printf("Error checking out %s in %s: %v\nstdout:\n%s\nstderr:\n%s\n", oldBranch, tpgbLocalPathOld, err, stdout, stderr)
+		os.Exit(1)
+	}
+
+	if err := updatePackageName("old", tpgbLocalPathOld, r); err != nil {
+		fmt.Printf("Error updating package name in %s: %v\n", tpgbLocalPathOld, err)
+		os.Exit(1)
+	}
+	if err := updatePackageName("new", tpgbLocalPath, r); err != nil {
+		fmt.Printf("Error updating package name in %s: %v\n", tpgbLocalPath, err)
+		os.Exit(1)
+	}
+
+	missingTestDetectorPath := filepath.Join(mmLocalPath, "tools", "missing-test-detector")
+	if stdout, stderr, err := r.Run(missingTestDetectorPath, "go", []string{"mod", "edit", "-replace", fmt.Sprintf("google/provider/%s=%s", "new", tpgbLocalPath)}, nil); err != nil {
+		fmt.Printf("Error running go mod edit: %v\nstdout:\n%s\nstderr:\n%s\n", err, stdout, stderr)
+	}
+	if stdout, stderr, err := r.Run(missingTestDetectorPath, "go", []string{"mod", "edit", "-replace", fmt.Sprintf("google/provider/%s=%s", "old", tpgbLocalPathOld)}, nil); err != nil {
+		fmt.Printf("Error running go mod edit: %v\nstdout:\n%s\nstderr:\n%s\n", err, stdout, stderr)
+	}
+	if stdout, stderr, err := r.Run(missingTestDetectorPath, "go", []string{"mod", "tidy"}, nil); err != nil {
+		fmt.Printf("Error running go mod tidy: %v\nstdout:\n%s\nstderr:\n%s\n", err, stdout, stderr)
+	}
+	missingTests, stderr, err := r.Run(missingTestDetectorPath, "go", []string{"run", ".", fmt.Sprintf("-services-dir=%s/google-beta/services", tpgbLocalPath)}, nil)
+	if err != nil {
+		fmt.Printf("Error running missing test detector: %v\nstdout:\n%s\nstderr:\n%s\n", err, missingTests, stderr)
+		missingTests = ""
+	} else {
+		fmt.Printf("Successfully ran missing test detector:\n%s\n", missingTests)
+	}
```

Comment:
I find this section pretty difficult to read. Could we factor these out into their own functions.

Maybe 

resultsBC := runBreakingChangeDetector
results mT := runMissingTestDetector

Reply:
Done